src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ValueListBuilder.cs

```diff
@@ -43,10 +43,26 @@ public int Length
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Append(T item)
         {
+            Span<T> span = _span;
             int pos = _pos;
-            if (pos >= _span.Length)
-                Grow();
+            if ((uint)pos < (uint)span.Length)
+            {
+                span[pos] = item;
+                _pos = pos + 1;
+            }
+            else
+            {
+                AddWithResize(item);
+            }
+        }
 
+        // Hide uncommon path
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void AddWithResize(T item)
```

Comment:
nit: move to Append as a nested function, it's a quite popular pattern in BCL

Reply:
I tried this initially but had issues with capturing the local state, and didn't want to have to pass `this`. Any suggestion?

Reply:
Yeah, this pattern doesn't work terribly well with ref structs.